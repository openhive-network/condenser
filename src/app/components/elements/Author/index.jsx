/* eslint react/prop-types: 0 */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router';
import tt from 'counterpart';
import { List } from 'immutable';
import { findDOMNode } from 'react-dom';
import Overlay from 'react-overlays/lib/Overlay';
// import shouldComponentUpdate from 'app//utils/shouldComponentUpdate';
import Icon from 'app//components/elements/Icon';
import Reputation from 'app//components/elements/Reputation';
import AffiliationMap from 'app//utils/AffiliationMap';
import UserTitle from 'app//components/elements/UserTitle';
import Sanitizer from 'app//utils/Sanitizer';
import AuthorDropdown from '../AuthorDropdown';

const { string, bool, number } = PropTypes;

const closers = [];

const fnCloseAll = () => {
    let close;
    while ((close = closers.shift())) {
        close();
    }
};

class Author extends React.Component {
    static propTypes = {
        author: string.isRequired,
        hideEditor: bool,
        follow: bool,
        mute: bool,
        authorRep: number,
        showAffiliation: bool,
        role: string,
        title: string,
        community: string,
        showRole: bool,
    };

    static defaultProps = {
        follow: true,
        mute: true,
        showAffiliation: false,
        role: '',
        title: '',
        community: '',
    };

    constructor(...args) {
        super(...args);
        this.state = { show: false };
        this.toggle = this.toggle.bind(this);
        this.close = this.close.bind(this);
    }

    componentDidMount() {
        if (!this.authorProfileLink) {
            return;
        }
        const node = findDOMNode(this.authorProfileLink);
        if (node.addEventListener) {
            node.addEventListener('click', this.toggle, false);
        } else {
            node.attachEvent('click', this.toggle, false);
        }
    }

    componentWillUnmount() {
        if (!this.authorProfileLink) {
            return;
        }
        const node = findDOMNode(this.authorProfileLink);
        if (node.removeEventListener) {
            node.removeEventListener('click', this.toggle);
        } else {
            node.detachEvent('click', this.toggle);
        }
    }

    toggle = (e) => {
        if (!(e.metaKey || e.ctrlKey)) {
            e.preventDefault();
            e.stopPropagation();
            const show = !this.state.show;
            fnCloseAll();
            if (show) {
                this.setState({ show });
                closers.push(this.close);
            }
        }
    };

    close = () => {
        this.setState({
            show: false,
        });
    };

    // shouldComponentUpdate = shouldComponentUpdate(this, 'Author');

    render() {
        const {
            author,
            authorRep,
            username,
            follow,
            mute,
            showAffiliation,
            blacklists,
            showRole,
            community,
            permlink,
            role,
            title,
        } = this.props;

        const warn = blacklists && (
            <span className="account_warn" title={blacklists.join(', ')}>
                (
                {blacklists.length}
                )
            </span>
        );

        const userTitle = (
            <span>
                {community && (
                    <UserTitle
                        username={username}
                        community={community}
                        author={author}
                        permlink={permlink}
                        role={showRole ? role : null}
                        title={title}
                        hideEdit={this.props.hideEditor}
                    />
                )}
                {showAffiliation && AffiliationMap[author] ? (
                    <span className="affiliation">{tt('g.affiliation_' + AffiliationMap[author])}</span>
                ) : null}
            </span>
        );

        if (!(follow || mute)) {
            return (
                <span className="author" itemProp="author" itemScope itemType="http://schema.org/Person">
                    <strong>
                        <Link to={'/@' + author}>{author}</Link>
                    </strong>
                    {' '}
                    <Reputation value={authorRep} />
                    {warn}
                    {userTitle}
                </span>
            );
        }

        return (
            <span className="Author">
                <span itemProp="author" itemScope itemType="http://schema.org/Person">
                    <strong>
                        <Link
                            ref={(link) => {
                                this.authorProfileLink = link;
                            }}
                            to={'/@' + author}
                        >
                            {author}
                            {' '}
                            <Reputation value={authorRep} />
                            <Icon name="dropdown-arrow" />
                        </Link>
                    </strong>
                    {warn}
                    {userTitle}
                </span>
                <Overlay
                    show={this.state.show}
                    onHide={this.close}
                    placement="bottom"
                    container={this}
                    target={() => findDOMNode(this.target)}
                    rootClose
                >
                    <AuthorDropdown
                        author={author}
                        follow={follow}
                        mute={mute}
                        authorRep={authorRep}
                        username={username}
                        blacklists={blacklists}
                    />
                </Overlay>
            </span>
        );
    }
}

export default connect((state, props) => {
    const { post, resolveCrossPost } = props;
    const blacklists = post.get('blacklists', List()).toJS();
    const crossPostedBy = post.get('cross_posted_by');

    let author = post.get('author');
    let authorRep = post.get('author_reputation');
    if (resolveCrossPost && crossPostedBy) {
        author = post.get('cross_post_author');
        authorRep = post.get('cross_post_author_reputation');
    }

    return {
        follow: typeof props.follow === 'undefined' ? true : props.follow,
        mute: typeof props.mute === 'undefined' ? props.follow : props.mute,
        username: state.user.getIn(['current', 'username']),
        authorRep,
        author,
        community: post.get('community'),
        permlink: post.get('permlink'),
        role: Sanitizer.getTextOnly(post.get('author_role')),
        title: Sanitizer.getTextOnly(post.get('author_title')),
        blacklists: blacklists.length > 0 ? blacklists : null,
        showRole: props.showRole,
    };
})(Author);
