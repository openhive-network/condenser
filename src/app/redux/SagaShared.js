import { fromJS } from 'immutable';
import {
 call, put, select, takeEvery, takeLatest
} from 'redux-saga/effects';
import { api } from '@hiveio/hive-js';
import { setUserPreferences } from 'app/utils/ServerApiClient';
import { callBridge } from 'app/utils/steemApi';
import * as globalActions from './GlobalReducer';
import * as appActions from './AppReducer';
import * as transactionActions from './TransactionReducer';

const wait = (ms) => new Promise((resolve) => {
        setTimeout(() => resolve(), ms);
    });

export const sharedWatches = [
    takeLatest(
        [appActions.SET_USER_PREFERENCES, appActions.TOGGLE_NIGHTMODE, appActions.TOGGLE_BLOGMODE],
        saveUserPreferences
    ),
    takeEvery('transaction/ERROR', showTransactionErrorNotification),
];

export function* getAccount(username, force = false) {
    let account = yield select((state) => state.global.get('accounts').get(username));

    // hive never serves `owner` prop (among others)
    const isLite = !!account && !account.get('owner');

    if (!account || force || isLite) {
        console.log('getAccount: loading', username, 'force?', force, 'lite?', isLite);

        [account] = yield call([api, api.getAccountsAsync], [username]);
        const accountWitness = yield call([api, api.callAsync], 'condenser_api.get_witness_by_account', [username]);

        if (account) {
            if (accountWitness) {
                account.account_is_witness = true;
            }

            account = fromJS(account);
            yield put(globalActions.receiveAccount({ account }));
        }
    }
    return account;
}

function* showTransactionErrorNotification() {
    const errors = yield select((state) => state.transaction.get('errors'));
    // eslint-disable-next-line no-restricted-syntax
    for (const [key, message] of errors) {
        // Do not display a notification for the bandwidthError key.
        if (key !== 'bandwidthError') {
            yield put(appActions.addNotification({ key, message }));
            yield put(transactionActions.deleteError({ key }));
        }
    }
}

export function* getContent({
 author, permlink, resolve, reject
}) {
    let content;
    while (!content) {
        console.log('getContent', author, permlink);
        content = yield call([api, api.getContentAsync], author, permlink);
        try {
            const converted = JSON.parse(content);
            if (converted.result && converted.result.length === 0) {
                content = null;
            }
        } catch (exception) {
            console.log('SagaShared::getContent()', exception.message);
        }

        if (content !== null && content.author == '') {
            // retry if content not found. #1870
            content = null;
            yield call(wait, 3000);
        }
    }

    //console.log('raw content> ', dbg(content));
    content = yield call(callBridge, 'normalize_post', { post: content });
    //console.log('normalized> ', dbg(content));

    yield put(globalActions.receiveContent({ content }));
    if (resolve && content) {
        resolve(content);
    } else if (reject && !content) {
        reject();
    }
}

/**
 * Save this user's preferences, either directly from the submitted payload or from whatever's saved in the store currently.
 *
 * @param {Object?} params.payload
 */
function* saveUserPreferences({ payload }) {
    if (payload) {
        yield setUserPreferences(payload);
    }

    const prefs = yield select((state) => state.app.get('user_preferences'));
    yield setUserPreferences(prefs.toJS());
}
