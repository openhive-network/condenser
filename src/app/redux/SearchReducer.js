import { Map, List, fromJS } from 'immutable';

const SEARCH_DISPATCH = 'search/SEARCH_DISPATCH';
const SEARCH_PENDING = 'search/SEARCH_PENDING';
const SEARCH_ERROR = 'search/SEARCH_ERROR';
const SEARCH_RESULT = 'search/SEARCH_RESULT';

const defaultSearchState = Map({
    pending: false,
    error: false,
    scrollId: false,
    result: List([]),
});

export default function reducer(state = defaultSearchState, action) {
    const { payload } = action;

    switch (action.type) {
        // Has a saga watcher.
        case SEARCH_DISPATCH: {
            return state;
        }
        case SEARCH_PENDING: {
            const { pending } = payload;
            state.setIn(['result'], undefined);
            state.setIn(['error'], undefined);
            return state.setIn(['pending'], pending);
        }
        case SEARCH_ERROR: {
            const { error } = payload;
            return state.setIn(['error'], error);
        }
        case SEARCH_RESULT: {
            const { results, scroll_id, append } = payload;

            if (results === null || results === undefined) return state;

            const posts = List(
                results.map(post => {
                    post.created = post.created_at;
                    post.author_rep = parseFloat(post.author_rep);
                    post.author_reputation = post.author_rep;
                    post.stats = { total_votes: post.total_votes };

                    return fromJS(post);
                })
            );

            let newState = {};
            if (!append) {
                newState = state.set('result', posts).set('scrollId', scroll_id);
            } else {
                // If append is true. need to process results and append them to previous result
                const updatedResults = state.get('result').concat(posts);
                newState = state.setIn(['result'], new List(updatedResults)).setIn(['scrollId'], scroll_id);
            }
            return newState;
        }
        default:
            return state;
    }
}

export const search = payload => ({
    type: SEARCH_DISPATCH,
    payload,
});
export const searchPending = payload => ({
    type: SEARCH_PENDING,
    payload,
});
export const searchError = payload => ({
    type: SEARCH_ERROR,
    payload,
});

export const searchResult = payload => ({
    type: SEARCH_RESULT,
    payload,
});
