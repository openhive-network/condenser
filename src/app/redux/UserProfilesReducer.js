import { fromJS } from 'immutable';

// Action constants
const ADD_USER_PROFILE = 'user_profile/ADD';
const ADD_LISTED_ACCOUNTS = 'user_profile/LISTED';
const ADD_HIVEBUZZ_BADGES = 'user_profile/HIVEBUZZ_BADGES';
const ADD_PEAKD_BADGES = 'user_profile/PEAKD_BADGES';
const SET_ERROR = 'user_profile/ERROR';

const defaultState = fromJS({
    profiles: {},
    error: false,
    hivebuzzBadges: [],
    peakdBadges: [],
});

export default function reducer(state = defaultState, action) {
    const { payload } = action;

    switch (action.type) {
        case ADD_USER_PROFILE: {
            if (payload) {
                return state.setIn(
                    ['profiles', payload.username],
                    fromJS(payload.account)
                );
            }
            return state;
        }

        case ADD_LISTED_ACCOUNTS: {
            if (payload) {
                return state.setIn(
                    ['listedAccounts', payload.username],
                    fromJS(payload.listed_accounts)
                );
            }
            return state;
        }

        case ADD_HIVEBUZZ_BADGES: {
            if (payload) {
                return state.setIn(
                    ['hivebuzzBadges', payload.username],
                    fromJS(
                        payload.badges
                            .filter(o => {
                                return o.state === 'on';
                            })
                            .map(o => {
                                return {
                                    id: o.id,
                                    type: o.type,
                                    name: o.name,
                                    title: o.title,
                                    url: o.url,
                                };
                            })
                    )
                );
            }
            return state;
        }

        case ADD_PEAKD_BADGES: {
            if (payload) {
                return state.setIn(
                    ['peakdBadges', payload.username],
                    fromJS(
                        payload.badges.map(o => {
                            return {
                                id: `${o.account}-${o.name}`,
                                type: 'badges',
                                name: o.name,
                                title: o.title,
                                url: `https://images.hive.blog/u/${
                                    o.name
                                }/avatar/small`,
                            };
                        })
                    )
                );
            }
            return state;
        }

        case SET_ERROR: {
            if (payload) {
                const { error } = payload;
                return state.setIn(['error'], error);
            }

            return state;
        }

        default:
            return state;
    }
}

// Action creators
export const addProfile = payload => ({
    type: ADD_USER_PROFILE,
    payload,
});

export const addHivebuzzBadges = payload => ({
    type: ADD_HIVEBUZZ_BADGES,
    payload,
});

export const addPeakdBadges = payload => ({
    type: ADD_PEAKD_BADGES,
    payload,
});

export const setError = payload => ({
    type: SET_ERROR,
    payload,
});

export const addList = payload => ({
    type: ADD_LISTED_ACCOUNTS,
    payload,
});
