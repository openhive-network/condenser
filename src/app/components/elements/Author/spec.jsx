import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import rootReducer from 'app/redux/RootReducer';
import { configure, shallow } from 'enzyme';

import Adapter from 'enzyme-adapter-react-16';
import Author from './index';

configure({ adapter: new Adapter() });

const store = createStore(rootReducer);

describe('<Author />', () => {
    const wrapper = shallow(
        <Provider store={store}>
            <Author />
        </Provider>
    );
    const container = wrapper.instance();

    it('renders without crashing', () => {
        expect(wrapper).toBeTruthy();
        expect(container).toBeTruthy();
        expect(wrapper).toMatchSnapshot();
    });
});
