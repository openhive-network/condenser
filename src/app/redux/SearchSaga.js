import { put, takeEvery, call } from 'redux-saga/effects';
import * as reducer from 'app/redux/SearchReducer';
import { conductSearch } from 'app/utils/ServerApiClient';

export const searchWatches = [takeEvery('search/SEARCH_DISPATCH', search)];

export function* search(action) {
    const { q, s, scroll_id } = action.payload;
    const append = !!scroll_id;
    yield put(reducer.searchPending({ pending: true }));
    try {
        const requestParams = {
            body: {
                q,
                sort: s ? s : 'relevance',
                scroll_id: scroll_id ? scroll_id : '',
            },
        };
        const searchResponse = yield call(conductSearch, requestParams);
        if (searchResponse.status >= 200 && searchResponse.status < 300) {
            const searchJSON = yield call([searchResponse, searchResponse.json]);
            yield put(reducer.searchResult({ ...searchJSON, append }));
        } else {
            yield put(reducer.searchError({ error: new Error('There was an error') }));
        }
    } catch (error) {
        yield put(reducer.searchError({ error }));
    }
    yield put(reducer.searchPending({ pending: false }));
}
