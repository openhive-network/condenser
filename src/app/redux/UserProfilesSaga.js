import { call, put, takeLatest } from 'redux-saga/effects';
import { callBridge, getHivePowerForUser } from 'app/utils/steemApi';
import HivebuzzApi from 'app/utils/hivebuzzApi';
import PeakdApi from 'app/utils/peakdApi';
import * as userProfileActions from './UserProfilesReducer';

const FETCH_PROFILE = 'userProfilesSaga/FETCH_PROFILE';
const FETCH_LISTS = 'userProfilesSaga/FETCH_LISTS';
const FETCH_HIVEBUZZ_BADGES = 'userProfileSaga/FETCH_HIVEBUZZ_BADGES';
const FETCH_PEAKD_BADGES = 'userProfileSaga/FETCH_PEAKD_BADGES';

export const userProfilesWatches = [
    takeLatest(FETCH_PROFILE, fetchUserProfile),
    takeLatest(FETCH_HIVEBUZZ_BADGES, fetchUserHivebuzzBadges),
    takeLatest(FETCH_PEAKD_BADGES, fetchUserPeakdBadges),
];

export function* fetchUserProfile(action) {
    const { account, observer } = action.payload;
    const ret = yield call(callBridge, 'get_profile', { account, observer });
    const hive_power = yield getHivePowerForUser(account);
    if (!ret) throw new Error('Account not found');
    yield put(
        userProfileActions.addProfile({
            username: account,
            account: { ...ret, stats: { ...ret.stats, sp: hive_power } },
        })
    );
}

export function* fetchUserHivebuzzBadges(action) {
    const { account } = action.payload;
    try {
        const ret = yield call(HivebuzzApi.getBadges, { account });
        if (!ret) {
            console.error('Hivebuzz badges error');
            yield put(
                userProfileActions.setError({
                    error: true,
                })
            );
        } else {
            yield put(
                userProfileActions.addHivebuzzBadges({
                    username: account,
                    badges: [...ret],
                })
            );
        }
    } catch (e) {
        console.error('Hivebuzz badges error');
        yield put(
            userProfileActions.setError({
                error: true,
            })
        );
    }
}

export function* fetchUserPeakdBadges(action) {
    const { account } = action.payload;
    try {
        const ret = yield call(PeakdApi.getBadges, { account });
        if (!ret) {
            console.error('Peakd badges error');
            yield put(
                userProfileActions.setError({
                    error: true,
                })
            );
        } else {
            yield put(
                userProfileActions.addPeakdBadges({
                    username: account,
                    badges: [...ret],
                })
            );
        }
    } catch (e) {
        console.error('Peakd badges error');
        yield put(
            userProfileActions.setError({
                error: true,
            })
        );
    }
}

// Action creators
export const actions = {
    fetchProfile: payload => ({
        type: FETCH_PROFILE,
        payload,
    }),
    fetchLists: payload => ({
        type: FETCH_LISTS,
        payload,
    }),
    fetchHivebuzzBadges: payload => ({
        type: FETCH_HIVEBUZZ_BADGES,
        payload,
    }),
    fetchPeakdBadges: payload => ({
        type: FETCH_PEAKD_BADGES,
        payload,
    }),
};
