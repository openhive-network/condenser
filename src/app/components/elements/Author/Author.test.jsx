import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import rootReducer from 'app/redux/RootReducer';
import { render } from '@testing-library/react';
import { fromJS } from 'immutable';

import Author from './index';
import post from '../../../../../api_mockdata/get_content';

const store = createStore(rootReducer);

describe('<Author />', () => {
    it('renders without crashing', () => {
        const { container, getByText } = render(
            <Provider store={store}>
                <Author
                    post={fromJS(post)}
                />
            </Provider>
        );
        expect(getByText('joeparys')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
