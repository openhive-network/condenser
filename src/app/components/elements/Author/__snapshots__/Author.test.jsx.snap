// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Author /> renders without crashing 1`] = `
<div>
  <span
    class="Author"
  >
    <span
      itemprop="author"
      itemscope=""
      itemtype="http://schema.org/Person"
    >
      <strong>
        <a>
          joeparys
           
          <span
            class="Reputation"
            title="missing translation: en.g.reputation"
          >
            (
            16463091699678
            )
          </span>
          <span
            class="Icon dropdown-arrow"
            style="display: inline-block; width: 1.12rem; height: 1.12rem;"
          >
            icon-mock
          </span>
        </a>
      </strong>
      <span />
    </span>
  </span>
</div>
`;
